### DISCUSSION TITLE
linter not recognising non-firedrake functions imported by firedrake wildcard import

### DISCUSSION THREAD

**Discussion by tom-new** [firedrakeproject/firedrake#2966] [Firedrake support] [answered]
I have firedrake succesfully installed on macOS 13.4 (Darwin). I can import firedrake without issue, and the basic demo `helmholtz.py` runs fine in terminal. VS Code's Python linter plug-in correctly recognises firedrake functions/classes (e.g. `UnitSquareMesh`), but does not seem to recognise many functions/classes/variables which I assume are from numpy (e.g. `grad`, `cos`, `pi`) but also others such as `SpatialCoordinate` are unrecognised, and I have no idea where they come from. Obviously the firedrake venv is activated in VS Code.

(On a related note, is there a reason why all the firedrake tutorial examples use wildcard imports? I was under the impression that they should be used extremely sparingly.)

**Comment by tom-new** [comment, marked as answer]
Solved. A second extension had a linting feature and was causing a conflict.

**Reply by connorjward** [reply]
I think the main reason why we use wildcard imports is because it makes the UFL expressions more elegant. Otherwise your equations would be littered with `fd.grad`, `fd.dx` etc or there would be a very hefty import line at the top of the script.
